commec/config/json_io.py

```diff
@@ -0,0 +1,453 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021-2024 International Biosecurity and Biosafety Initiative for Science
+'''
+   Set of tools for retrieving and storing information important to screen
+    outputs. Information is stored as a structure of dataclasses, and are
+    converted between the dataclass / dict / json_file as required. The
+    conversions are done dynamically, and it is recommended to only use and
+    interact with the dataclasses only, to maintain version format, and not
+    create erroneous outputs to the JSON which wont be read back in. This
+    ensures an expected i/o behaviour.
+
+    The single exception to this is the "annotations" dictionary, present
+    in the HitDescription, which contains non-structured information, and is
+    populated with differing information under differing keys depending on
+    which step the information is derived (Biorisk, Taxonomy etc)
+
+    In this way, the JSON object serves as a common state, that can be updated
+    whilst not being temporally appended like a log file i.e. .screen file.
+
+    The JSON stores all pertinent information of a run.
+'''
+
+# Consider whether this can get away with being part of config. rename to IO config?
+
+import json
+import string
+import os
+from dataclasses import dataclass, asdict, fields, field, is_dataclass
+from typing import Dict, Type, get_origin, Any, get_args, List, Iterator, Tuple
+from enum import StrEnum
+from commec.tools.search_handler import SearchToolVersion
+
+# Seperate versioning for the output JSON.
+JSON_COMMEC_FORMAT_VERSION = "0.1"
+
+def guess_domain(search_string : str) -> str:
+    """ 
+    Given a string description, try to determine 
+    which domain of life this has come from. Temporary work around
+    until we can retrieve this data directly from biorisk outputs.
+    """
+    def contains(search_string : str, search_terms):
+        for token in search_terms:
+            if search_string.find(token) == -1:
+                continue
+            return True
+        return False
+
+    search_token = search_string.lower()
+    if contains(search_token, ["vir", "capsid", "RNA Polymerase"]):
+        return "Virus"
+    if contains(search_token, ["cillus","bact","coccus","phila","ella","cocci","coli"]):
+        return "Bacteria"
+    if contains(search_token, ["eukary","nucleus","sona","odium","myces"]):
+        return "Eukaryote"
+    return "not assigned"
+
+class CommecScreenStep(StrEnum):
```

Comment:
Restructuring proposal (related to the `config` --> `io`) that I proposed before; I think we should separate out the enums / bits holding the actual progress of screening from the JSON serialization logic.

What if we did `screeners` --> `screening` and moved some of these data classes in there?